Implementing edit title

zsolt270 opened the pull request:
Finished the edit-title feature. I tested the function with the tests that I wrote and manually, both of them were successful. I would ask u guys to also check the new feature and write a review, also merge if its mergeable.

zsolt270 committed:
finished the edit title params validation

zsolt270 committed:
changed the returned value to array in the validation function

zsolt270 committed:
added the edit title function with checking if the given id is found

zsolt270 committed:
finished the edit title function only the test remains

zsolt270 committed:
finished the test for the edit-title function. Only the validation test remains.

zsolt270 committed:
finished the validation test of the feature

zsolt270 committed:
finished the edit-title feature and added comments to the function

Intcognito commented on src/app.js:
Using JSDoc to describe the function is a very good practice!!
sidenote: the user story described the command as: "update-title", but "edit-title" is understandable as well.

Intcognito commented on src/todo.js:
Using the already implemented findById method is a good way to avoid later refactoring!

Intcognito commented on src/todo.spec.js:
Test is covering both the returned object and if the list is updated

Intcognito commented on src/validate.js:
In line 36 you check if the newTitle is a string and at least 1 character long:
36 - if (!isNaN(newTitle) || newTitle.length < 1)

I think if the newTitle is "undefined", than the "!isNaN(newTitle)" expression will return false (and the error won't be thrown), as undefined is not a number, but also not a string.

I would consider checking the type: 
if (typeof newTitle !== 'string' || newTitle.length < 1)

Maybe the code works as intended, but please check it! Thanks!

zsolt270 replied:
I think it should be fine. The only case I can think of when the newTitle is undifined is when the user doesnt give a second parameter. In that case the first if check will throw an Error, so the newTitle cant be undefined, but if there are other cases then please let me know and I will check them.

Also at first I tried with the solution u mentioned:
if (typeof newTitle !== 'string' || newTitle.length < 1)
but it would have few problems in my opinion.

For Example:
-the first condition (typeof newTitle !== 'string') would never trigger because I assume that the parameters that come from the command line are always strings. (I checked the first parameter type which is the id(a number) and it gave me string as a result. That is the reason why I think this.)

-Also for example if I would try to paseInt() the newTitle before the if statement to check if it gives back a number or NaN. Then it would give back number as a result for titles like theese: "40 push-ups", "20 sit-ups". So basically for every title that starts with a number.

So I think checking if it is !isNaN(newTitle) should be fine, but if it isnt maybe I could just add another check to it for undefined parameter type. ( || typeof newTitle === undefined or something like that).

Intcognito replied:
You are right. I also couldn't figure out a scenario when the paramters given don't throw an error and not a string. It just seemed wrong, but your reasoning is sound. Since any input given will be in string format, this typeof check wouldn't catch any errors. In the other hand if no param is given, than the other cases will cath it!
Thanks for checking it!

Intcognito commented on src/validate.spec.js:
validation tests cover plenty of cases

Intcognito committed:
Merge branch 'main' into implementing-edit-title

Intcognito committed:
Update validate.spec.js

}) - closing tag was lost on merge

Intcognito committed:
Update todo.spec.js

}); - closing tag was lost on merge

Intcognito committed:
Update todo.spec.js

fixing });

Intcognito committed:
Update validate.spec.js

}); issue

Intcognito committed:
Update validate.spec.js

Intcognito committed:
Update todo.spec.js

Intcognito committed:
Update todo.spec.js

reformatted test file for github actions isn't able to run it

Intcognito committed:
Commented out a test case

I removed a test case which is cheking if the todos list is updated after title changed. Maybe that is causing jest to throw an error.

Intcognito committed:
Commented out the whole "edit-title" test

To see if jest can run without this part

Intcognito committed:
Copy-pasted a working version without new test cases

I saved the new test cases in a separate file and copy-pasted a previous (supposedly working) version of this test file to see where is the error.
Maybe the error is not in this file, just the error message is misleading.
